src/hotspot/share/opto/parseHelper.cpp

```diff
@@ -827,24 +818,133 @@ void AllocationStateMerger::merge(const PEAState& newin, GraphKit* kit, RegionNo
+// been materialized before. the phi is 'reducible Object-Phi' in JDK-828706 and the original object is scalar replaceable!
+//
+// obj' = PHI(Region, OriginalObj, ClonedObj)
+// and OriginalObj is NonEscape but NSR; CloendObj is Global/ArgEscape
+//
+// JDK-8287061 transforms it to =>
+// obj' = PHI(Region, null, ClonedObj)
+// selector = PHI(Region, 0, 1)
+//
+// since OriginalObj is NonEscape, it is replaced by scalars.
+//
+static Node* ensure_object_materialized(Node* var, PEAState& state, SafePointNode* from_map, RegionNode* r, int pnum) {
+  // skip passive materialize for time being.
+  // if JDK-8287061 can guarantee to replace the orignial allocation, we don't need to worry about partial redundancy.
+  return var;
+}
+
+// Merge phi node incrementally.
+// we check all merged inputs in _state.
+// 1. all inputs refer to the same ObjID, then phi is created as alias of ObjID
+// 2. otherwise, any input is alias with a 'virtual' object needs to convert to 'Escaped'. replace input with merged_value.
+// 3. otherwise, if any input is aliased with an Escaped object. replace input with merged value.
+void AllocationStateMerger::process_phi(PhiNode* phi, GraphKit* kit, RegionNode* region, int pnum) {
+  ObjID unique = nullptr;
+  bool materialized = false;
+  bool same_obj = true;
+  PartialEscapeAnalysis* pea = kit->PEA();
+
+  if (pea == nullptr) return;
+
+  for (uint i = 1; i < phi->req(); ++i) {
+    if (region->in(i) == nullptr || region->in(i)->is_top())
+      continue;
+
+    Node* node = phi->in(i);
+    ObjID obj = pea->is_alias(node);
+    if (obj != nullptr) {
+      if (unique == nullptr) {
+        unique = obj;
+      } else if (unique != obj) {
+        same_obj = false;
+      }
+      EscapedState* es = _state.as_escaped(pea, node);
+      if (es != nullptr) {
+        materialized |= es->has_materialized();
+      }
+    } else {
+      same_obj = false;
+    }
+  }
+
+  if (same_obj) {
+    //xliu: should I also check pnum == 1?
+    // phi nodes for exception handler may have leave normal paths vacant.
+    pea->add_alias(unique, phi);
+  } else {
+    bool printed = false;
+
+    for (uint i = 1; i < phi->req(); ++i) {
+      if (region->in(i) == nullptr || region->in(i)->is_top())
+        continue;
+
+      Node* node = phi->in(i);
+      ObjID obj = pea->is_alias(node);
+      if (obj != nullptr && _state.contains(obj)) {
+        ObjectState* os = _state.get_object_state(obj);
+        if (os->is_virtual()) {
+          Node* n = ensure_object_materialized(node, _state, kit->map(), region, pnum);
+          os = _state.escape(obj, n, materialized);
+        }
+        EscapedState* es = static_cast<EscapedState*>(os);
+        Node* value = es->merged_value();
+        // if value is an alias of obj, do nothing.
+        // or we select the value from i-th predecessor.
+        if (value->is_Phi() && pea->is_alias(value) != obj) {
```

Comment:
I'm having trouble understanding this part. I think what we want to do is, if we are newly escaping `node`, we should update the phi node. But I'm not sure what this if statement is trying to check.

Reply:
I change it a little bit. What we need to avoid the situation that "value = es->merged_value()" is phi. if value is phi and it's legit for current map, I select its i-th input including phi node. if value is phi, node != value. I check it using assertion.  


I think it's more straightforward. what do you think? 
```java
        Node* value = es->merged_value();
        if (value->is_Phi() && value->in(0) == region) {
          value = value->in(i);
        }

        if (node != value) {
          assert(value != phi, "sanity");
```

Reply:
Yeah, it makes sense to me now